File: day9/python/day9.py
from re import compile

marker = compile(r'^\((\d+)x(\d+)\)')

def decompress(line, rec=False):
    if line == "":
        return 0

    match = marker.match(line)
    if match:
        length, times = map(int, match.groups())
        remainder = line[len(match.group(0)):]
        x = decompress(remainder[:length], rec) if rec else length
        return x * times + decompress(remainder[length:], rec)
    else:
        return 1 + decompress(line[1:])

def day9(input):
    return decompress(input), decompress(input, True)

input = open("../input.txt").read()
print(day9(input))

File: day09/haskell/day9.hs
import Control.Arrow ((***))
import Data.List (break)

marker :: String -> (Maybe (Int, Int), String)
marker ('(':xs) = (Just . parse *** tail) . break (==')') $ xs
    where parse = (read *** read . tail) . break (=='x')
marker s = (Nothing, s)

decompress :: Bool -> String -> Int
decompress b = go b . marker
    where go _ (_, []) = 0
          go _ (Nothing, s) = 1 + rec tail s
          go False (Just (l, t), s) = l * t + rec (drop l) s
          go True (Just (l, t), s) = rec (take l) s * t + rec (drop l) s
          rec f = decompress b . f

main :: IO ()
main = do
    input <- readFile "../input.txt"
    print . decompress False $ input
    print . decompress True $ input